### DISCUSSION TITLE
Starting to support pypy3.11, need guidance for a `cfg` stanza

### DISCUSSION THREAD

**Discussion by mattip** [PyO3/pyo3#4755] [Questions] [answered]
Over at [a binary build repo](https://github.com/pypy/binary-testing) I am trying out PyPy3.11 (PyPy with python3.11) on a [branch off my fork](https://github.com/PyO3/pyo3/compare/main...mattip:pyo3:pypy3.11) of PyO3. I changed the C interface for `PyObject_DelAttr` to be more like CPython3.11: it is a macro that uses `PyObject_SetAttr`. How can I express that in this `cfg` stanza? Do I need to define `PyPy_3_11` somewhere? https://github.com/PyO3/pyo3/blob/6ec499b2bf266b207e16ccacbc6dc8ab4929bd54/pyo3-ffi/src/abstract_.rs#L12-L15

**Comment by davidhewitt** [comment, marked as answer]
We define our cfgs so that `Py_3_11` will be set for Python versions 3.11 _and up_, which will include PyPy if we detect that version.

So to get "compiling for PyPy with at least a 3.11 interpreter", you can use `cfg(all(PyPy, Py_3_11))`.

So for this full clause, you'd want to change it to something like

```rust
#[cfg(
    all(
        not(Py_3_13), // CPython exposed as a function in 3.13, in object.h 
        not(all(PyPy, not(Py_3_11))) // PyPy exposed as a function until PyPy 3.10, used macro in 3.11+
    )
)]
pub unsafe fn PyObject_DelAttr(o: *mut PyObject, attr_name: *mut PyObject) -> c_int { 
    PyObject_SetAttr(o, attr_name, std::ptr::null_mut()) 
} 
```

... amusing that CPython went the other direction in 3.13. I guess eventually when PyPy supports 3.13 you'll do similar and revert to the existing function form, to keep in sync with CPython?

**Reply by davidhewitt** [reply]
It might also be easier to reason about if you swap some `all` for `any` and invert the `not`s, as per usual boolean logic rules. Not sure in this case, looks like it'll be complex either way. The extern symbol in `object.rs` will need the inverted clause (probably just wrap whatever you settle on in an enclosing `not()`).

**Comment by mattip** [comment]
> We define our cfgs so that Py_3_11 will be set for Python versions 3.11 and up, which will include PyPy if we detect that version.

Great, and thanks for the other hints.
